From 28bee6e3ff5c62f74d242f47e7cb26e3b6fc2ea8 Mon Sep 17 00:00:00 2001
From: Voon Weifeng <weifeng.voon@intel.com>
Date: Mon, 9 Dec 2019 16:02:57 +0800
Subject: [PATCH 06/17] net: stmmac: add TGL ingress and egress phy latency

Added Marvell PHY 88E2110 path latency for TGL platform data to reduce
the link propagation delay during gPTP clock sync.

Signed-off-by: Voon Weifeng <weifeng.voon@intel.com>
---
 drivers/net/ethernet/stmicro/stmmac/stmmac_main.c |  6 ++++++
 drivers/net/ethernet/stmicro/stmmac/stmmac_pci.c  | 10 ++++++++++
 include/linux/stmmac.h                            |  2 ++
 3 files changed, 18 insertions(+)

diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
index 9b7719500151..d62618249f01 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_main.c
@@ -458,6 +458,9 @@ static void stmmac_get_tx_hwtstamp(struct stmmac_priv *priv,
 		case SPEED_1000:
 			adjust = priv->plat->phy_tx_latency_1000;
 			break;
+		case SPEED_2500:
+			adjust = priv->plat->phy_tx_latency_2500;
+			break;
 		}
 
 		ns += adjust;
@@ -507,6 +510,9 @@ void stmmac_get_rx_hwtstamp(struct stmmac_priv *priv, struct dma_desc *p,
 		case SPEED_1000:
 			adjust = priv->plat->phy_rx_latency_1000;
 			break;
+		case SPEED_2500:
+			adjust = priv->plat->phy_rx_latency_2500;
+			break;
 		}
 
 		ns -= adjust;
diff --git a/drivers/net/ethernet/stmicro/stmmac/stmmac_pci.c b/drivers/net/ethernet/stmicro/stmmac/stmmac_pci.c
index 73fd820ee348..a38ef4b4180d 100644
--- a/drivers/net/ethernet/stmicro/stmmac/stmmac_pci.c
+++ b/drivers/net/ethernet/stmicro/stmmac/stmmac_pci.c
@@ -464,6 +464,16 @@ static int tgl_common_data(struct pci_dev *pdev,
 	/* Maximum TX XDP queue */
 	plat->max_combined = 2;
 
+	/* TX and RX Marvell 88E2110 PHY latency (ns) */
+	plat->phy_tx_latency_10 = 6652;
+	plat->phy_tx_latency_100 = 1152;
+	plat->phy_tx_latency_1000 = 297;
+	plat->phy_tx_latency_2500 = 2772;
+	plat->phy_rx_latency_10 = 12490;
+	plat->phy_rx_latency_100 = 1472;
+	plat->phy_rx_latency_1000 = 405;
+	plat->phy_rx_latency_2500 = 2638;
+
 	plat->clk_ptp_rate = 200000000;
 	ret = intel_mgbe_common_data(pdev, plat);
 	if (ret)
diff --git a/include/linux/stmmac.h b/include/linux/stmmac.h
index 9fcaff572cb7..bde0a10a4eba 100644
--- a/include/linux/stmmac.h
+++ b/include/linux/stmmac.h
@@ -235,9 +235,11 @@ struct plat_stmmacenet_data {
 	bool is_hfpga;
 	bool ehl_ao_wa;
 	/* TX and RX PHY latency (ns) */
+	u64 phy_tx_latency_2500;
 	u64 phy_tx_latency_1000;
 	u64 phy_tx_latency_100;
 	u64 phy_tx_latency_10;
+	u64 phy_rx_latency_2500;
 	u64 phy_rx_latency_1000;
 	u64 phy_rx_latency_100;
 	u64 phy_rx_latency_10;
-- 
1.9.1

